contracts/exchange/Exchange_ERC20_1155.cairo

```diff
@@ -0,0 +1,552 @@
+    let (token_reserves) = IERC1155.balanceOf(token_address_, contract, token_id)
+
+    # Take the token amount
+    IERC1155.safeTransferFrom(token_address_, caller, contract, token_id, token_amount.low)
+
+    # Calculate prices
+    let (currency_amount) = get_sell_price(token_amount, currency_reserves_, Uint256(token_reserves, 0))
+
+    # Check min_currency_amount
+    let (above_min_curr) = uint256_le(min_currency_amount, currency_amount)
+    assert_not_zero(above_min_curr)
+
+    # Transfer currency
+    IERC20.transfer(currency_address_, caller, currency_amount)
+    tempvar syscall_ptr :felt* = syscall_ptr
+
+    # Update reserves
+    let (new_reserves) = uint256_sub(currency_reserves_, currency_amount)
+    currency_reserves.write(token_id, new_reserves)
+
+    # Emit event
+    currency_purchased.emit(caller, currency_amount, token_id, token_amount)
+
+    return (currency_amount)
+end
+
+
+#
+# Pricing
+#
+
+
+@view
+func get_buy_price {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        token_amount: Uint256,
+        currency_reserves: Uint256,
+        token_reserves: Uint256,
+    ) -> (
+        price: Uint256
+    ):
+    alloc_locals
+    let (lp_fee_thousands_) = lp_fee_thousands.read()
+    let (lp_fee) = uint256_sub(Uint256(1000, 0), lp_fee_thousands_)
+
+    # Calculate price
+    let (numerator, mul_overflow) = uint256_mul(currency_reserves, token_amount)
+    assert mul_overflow = Uint256(0, 0)
+    let (denominator) = uint256_sub(token_reserves, token_amount)
+
+    # Add LP fee
+    let (numerator, mul_overflow) = uint256_mul(numerator, Uint256(1000, 0))
+    assert mul_overflow = Uint256(0, 0)
+    let (denominator, mul_overflow) = uint256_mul(denominator, lp_fee)
+    assert mul_overflow = Uint256(0, 0)
+
+    let (price, remainder) = uint256_unsigned_div_rem(numerator, denominator)
+
+    let (is_z) = uint256_eq(remainder, Uint256(0, 0))
+    if is_z == (1):
+        return (price)
+    end
+
+    let (price, _) = uint256_add(price, Uint256(1, 0))
+    return (price)
+
+end
+
+
+@view
+func get_sell_price {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        token_amount: Uint256,
+        currency_reserves: Uint256,
+        token_reserves: Uint256,
+    ) -> (
+        price: Uint256
+    ):
+    alloc_locals
+    let (lp_fee_thousands_) = lp_fee_thousands.read()
+    let (lp_fee) = uint256_sub(Uint256(1000, 0), lp_fee_thousands_)
+
+    # Calculate price
+    let (numerator, mul_overflow) = uint256_mul(token_amount, currency_reserves)
+    assert mul_overflow = Uint256(0, 0)
+    let (denominator, is_overflow) = uint256_add(token_reserves, token_amount)
+    assert is_overflow = 0
+
+    # Add LP fee
+    let (numerator, mul_overflow) = uint256_mul(numerator, lp_fee)
+    assert mul_overflow = Uint256(0, 0)
+    let (denominator, mul_overflow) = uint256_mul(denominator, Uint256(1000, 0))
```

Comment:
The [way the denominator is calculated](https://github.com/0xsequence/niftyswap/blob/master/src/contracts/exchange/NiftyswapExchange20.sol#L311) in `NiftySwap` is different. This implementation does `(token_reserves + token_amount) * 1000` but NS does `token_reserves * 1000 + token_amount_with_fees`